<!-- LOCAL -->
<launch>
    <!-- WARNING: ROS2 does NOT scope launch arguments. Make sure they have unique names -->
    <arg name="local_odometry_in_topic" default="/$(env ROBOT_NAME)/odometry_conversion/odometry" />
    <arg name="local_disparity_in_topic" default="/$(env ROBOT_NAME)/sensors/front_stereo/disparity" />
    <!-- <arg name="local_disparity_in_topic" default="/fake_disparity" /> -->
    <arg name="local_depth_in_topic" default="/$(env ROBOT_NAME)/sensors/front_stereo/left/depth" />

    <log message="local_odometry_in_topic has $(var local_odometry_in_topic)"/>

    <!-- customize takeoff_landing_planner.launch.xml -->
    <node pkg="takeoff_landing_planner"
        namespace="takeoff_landing_planner"
        exec="takeoff_landing_planner" output="screen">
        <param from="$(find-pkg-share takeoff_landing_planner)/config/takeoff_landing_planner.yaml"
            allow_substs="true" />

        <!-- remap glue to AirStack -->
        <remap from="trajectory_override"
            to="/$(env ROBOT_NAME)/trajectory_controller/trajectory_override" />
        <remap from="trajectory_completion_percentage"
            to="/$(env ROBOT_NAME)/trajectory_controller/trajectory_completion_percentage" />
        <remap from="tracking_point"
            to="/$(env ROBOT_NAME)/trajectory_controller/tracking_point" />
        <remap from="odometry" to="$(var local_odometry_in_topic)" />
        <remap from="mavros/cmd/takeoff" to="/$(env ROBOT_NAME)/interface/mavros/cmd/takeoff" />
    </node>

    <!-- Control -->
    <!-- customize trajectory_controller.launch.xml -->
    <node pkg="trajectory_controller" exec="trajectory_controller"
        namespace="trajectory_controller"
        output="screen">
        <param name="tf_prefix" value="" />
        <param name="target_frame" value="map" />
        <param name="tracking_point_distance_limit" value="1000.5" />
        <param name="velocity_look_ahead_time" value="0.9" />
        <!-- look ahead time controls the speed, greater is faster -->
        <param name="look_ahead_time" value="1.0" />
        <param name="virtual_tracking_ahead_time" value="0.5" />
        <param name="min_virtual_tracking_velocity" value="0.5" />
        <param name="sphere_radius" value="1.0" />
        <param name="ff_min_velocity" value="0." />
        <param name="search_ahead_factor" value="1.5" />
        <param name="transition_velocity_scale" value="1.0" />
        <param name="traj_vis_thickness" value="0.03" />
        <param name="rewind_skip_max_velocity" value="0.1" />
        <param name="rewind_skip_max_distance" value="0.1" />
        <param name="velocity_sphere_radius_multiplier" value="0.5" />

        <!-- remap glue to AirStack -->
        <remap from="odometry" to="$(var local_odometry_in_topic)" />

    </node>

    <node pkg="pid_controller" exec="pid_controller" namespace="control" output="screen">
      <param name="target_frame" value="base_link_stabilized" />

      <param name="x_p"         value="1.0" />
      <param name="x_i"         value="0.0" />
      <param name="x_d"         value="0.0" />
      <param name="x_ff"        value="0.0" />
      <param name="x_d_alpha"   value="0.0" />
      <param name="x_min"       value="-3.0" />
      <param name="x_max"       value="3.0" />
      <param name="x_constant"  value="0.0" />
      
      <param name="y_p"         value="1.0" />
      <param name="y_i"         value="0.0" />
      <param name="y_d"         value="0.0" />
      <param name="y_ff"        value="0.0" />
      <param name="y_d_alpha"   value="0.0" />
      <param name="y_min"       value="-3.0" />
      <param name="y_max"       value="3.0" />
      <param name="y_constant"  value="0.0" />
      
      <param name="z_p"         value="1.0" />
      <param name="z_i"         value="0.0" />
      <param name="z_d"         value="0.0" />
      <param name="z_ff"        value="0.0" />
      <param name="z_d_alpha"   value="0.0" />
      <param name="z_min"       value="-1.0" />
      <param name="z_max"       value="1.0" />
      <param name="z_constant"  value="0.0" />
      
      <param name="vx_p"         value="0.2" />
      <param name="vx_i"         value="0.025" />
      <param name="vx_d"         value="0.0" />
      <param name="vx_ff"        value="0.0" />
      <param name="vx_d_alpha"   value="0.9" />
      <param name="vx_min"       value="-0.34" />
      <param name="vx_max"       value="0.34" />
      <param name="vx_constant"  value="0.0" />
      
      <param name="vy_p"         value="0.2" />
      <param name="vy_i"         value="0.025" />
      <param name="vy_d"         value="0.0" />
      <param name="vy_ff"        value="0.0" />
      <param name="vy_d_alpha"   value="0.9" />
      <param name="vy_min"       value="-0.34" />
      <param name="vy_max"       value="0.34" />
      <param name="vy_constant"  value="0.0" />
      
      <param name="vz_p"         value="0.2" />
      <param name="vz_i"         value="0.1" />
      <param name="vz_d"         value="0.0" />
      <param name="vz_ff"        value="0.0" />
      <param name="vz_d_alpha"   value="0.0" />
      <param name="vz_min"       value="-0.5" />
      <param name="vz_max"       value="1.0" />
      <param name="vz_constant"  value="0.5" />
      
      <!-- remap glue to AirStack -->
      <remap from="odometry" to="$(var local_odometry_in_topic)" />
      <remap from="tracking_point" to="/$(env ROBOT_NAME)/trajectory_controller/tracking_point" />
      <remap from="command" to="/$(env ROBOT_NAME)/interface/cmd_roll_pitch_yawrate_thrust" />
    </node>

</launch>
